recipes/opentelemetry-cpp/all/conanfile.py

```diff
@@ -0,0 +1,128 @@
+    settings = "os", "arch", "compiler", "build_type"
+    options = {
+        "fPIC": [True, False],
+        "shared": [True, False],
+    }
+    default_options = {
+        "fPIC": True,
+        "shared": False,
+    }
+    exports_sources = "CMakeLists.txt"
+
+    short_paths = True
+    _cmake = None
+
+    def validate(self):
+        if self.settings.arch != "x86_64":
+            raise ConanInvalidConfiguration("Architecture not supported")
+
+        if (self.settings.compiler == "Visual Studio" and
+           tools.Version(self.settings.compiler.version) < "16"):
+            raise ConanInvalidConfiguration("Visual Studio 2019 or higher required")
+
+        if self.settings.os != "Linux" and self.options.shared:
+            raise ConanInvalidConfiguration("Building shared libraries is only supported on Linux")
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    def source(self):
+        tools.get(
+            **self.conan_data["sources"][self.version],
+            destination=self._source_subfolder,
+            strip_root=True)
+
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+
+        self._cmake = CMake(self)
+        defs = {
+          "WITH_ABSEIL": True,
+          "WITH_OTLP": True,
+          "WITH_JAEGER": True,
+          "WITH_EXAMPLES": False,
+          "BUILD_TESTING": False
+        }
+        self._cmake.configure(defs=defs, build_folder=self._build_subfolder)
+        return self._cmake
+
+    def _patch_sources(self):
+        protos_path = self.deps_cpp_info["opentelemetry-proto"].res_paths[0].replace("\\", "/")
+        protos_cmake_path = os.path.join(
+            self._source_subfolder,
+            "cmake",
+            "opentelemetry-proto.cmake")
+        tools.replace_in_file(
+            protos_cmake_path,
+            "set(PROTO_PATH \"${CMAKE_CURRENT_SOURCE_DIR}/third_party/opentelemetry-proto\")",
+            f"set(PROTO_PATH \"{protos_path}\")")
+        tools.rmdir(os.path.join(self._source_subfolder, "api", "include", "opentelemetry", "nostd", "absl"))
+
+    def build(self):
+        self._patch_sources()
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def package(self):
+        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+
+    def package_info(self):
+        if self.settings.os in ("Linux", "FreeBSD"):
+            self.cpp_info.system_libs = ["pthread"]
+
+        self.cpp_info.libs = [
```

Comment:
Could you elaborate?

Reply:
The library allows its users to only use select libraries.
e.g. only use [`opentelemetry-cpp::opentelemetry_version`](https://github.com/open-telemetry/opentelemetry-cpp/blob/d1f0b136e3cf88e3d13b2b979520fafe32705319/sdk/src/version/CMakeLists.txt#L12-L17) instead of needing to link to all libraries provided by opentelementry-cpp.
By using conan modules/components, this information can be provided.

When not providing this information, some dependencies might fail to build when these targets are not available.

Reply:
It does, although using one lib of it is quite useless. I don't think anyone would do it with Otel C++. Don't really know why they split it up like that.

Reply:
@madebr e.g. see this section: https://github.com/open-telemetry/opentelemetry-cpp/blob/main/INSTALL.md#incorporating-into-an-existing-cmake-project

Reply:
I was referring to the sentence `As of now, this will import targets for both API and SDK`. Right now, only one target is created.
Also, should the test package test using `OPENTELEMETRY_CPP_INCLUDE_DIRS` and `OPENTELEMETRY_CPP_LIBRARIES` ?

Reply:
I'm not sure, isn't the opentelemetry-cpp target created by Conan here?

Reply:
Conan created `opentelemetry-cpp::opentelemetry-cpp` by default.
We aim with cci recipes to mimic the upstream cmake scripts as best as we can.
In this case, upstream creates cmake targets for each library. 
So upstream is way more detailed then the conan default.

Reply:
Maybe, but no one is using these libs separately 🤷

Reply:
Okay then.
Then make sure that `OPENTELEMETRY_CPP_INCLUDE_DIRS` and `OPENTELEMETRY_CPP_LIBRARIES ` are working and tested by the test package.
It looks like those are used by splunk-opentelemetry of #7565.
Also, please add a FIXME about missing modules to `package_info`.

Reply:
I'm kinda lost at package_info how to achieve it, if the target name for find_package is `opentelemetry-cpp`, how can I also make Conan (1.42) export `OPENTELEMETRY_CPP_INCLUDE_DIRS` and `OPENTELEMETRY_CPP_LIBRARIES`?

Reply:
The easiest way to achieve this is to create a `.cmake` file in the `package` method that sets these cmake variables, using the target(s) created by cmake_find_package.
Something like
```python
tools.save(os.path.join(self.package_folder, self._module_file_rel_path), textwrap.dedent("""\
    set(OPENTELEMETRY_CPP_INCLUDE_DIRS )
    set(OPENTELEMETRY_CPP_LIBRARIES opentelemetry-cpp::opentelemetry-cpp)
"""))
```
And make sure to add `self._module_file_rel_path` to build_modules and `self._module_subfolder` to builddirs.

I suggest grepping for some names in the cci repo for examples on how other recipes do this.

Reply:
@madebr the splunk-opentelemetry PR will get updated after this one, `find_package` works with `opentelemetry-cpp_LIBRARIES`, but how can I get it to export `OPENTELEMETRY_CPP_LIBRARIES` while keeping the package name (for find_package) `opentelemetry-cpp`?

Reply:
See https://github.com/conan-io/conan-center-index/pull/7799#discussion_r739723107
Write a custom `.cmake` script to the package folder in the `package` method + add the dir to builddirs + add the cmake script to build_modules.
See [the docs](https://docs.conan.io/en/latest/reference/conanfile/attributes.html?highlight=build_modules#cpp-info) + other recipes.

Check e.g. the openssl/3.0, xapian-core recipes on how this can be done.

Reply:
@madebr Thanks for the help, I've moved the recipe to use components and added a variables file.